src/issue.rs

```diff
@@ -31,82 +30,525 @@ fn issue_actor_and_project(
+	if let Some(prev_project) = db_entry.last_confirmed_project {
+		db_entry.issue_project_state =
+			Some(ProjectState::Confirmed(prev_project));
+		// reattach the last confirmed project
+		github_bot.create_project_card(
+			prev_project,
+			issue_id,
+			github::ProjectCardContentType::Issue,
+		)?;
+	} else {
+		db_entry.issue_project_state = None;
+	}
+	if let Some(matrix_id) = github_to_matrix
+		.get(&actor.login)
+		.and_then(|matrix_id| matrix::parse_id(matrix_id))
+	{
+		matrix_bot.send_private_message(
+			&matrix_id,
+			&ISSUE_REVERT_PROJECT_NOTIFICATION.replace("{1}", &issue_html_url),
+		)?;
+	} else {
+		// no matrix id to message
+	}
+	Ok(DbEntryState::Update)
+}
+
+fn author_non_admin_project_state_confirmed(
+	db_entry: &mut DbEntry,
+	matrix_bot: &MatrixBot,
+	issue: &github::Issue,
+	_default_channel_id: &str,
+	project: &github::Project,
+	project_column: &github::ProjectColumn,
+	project_info: &project_info::ProjectInfo,
+	confirmed_id: i64,
+) -> Result<DbEntryState> {
+	let issue_id = issue.id.context(error::MissingData)?;
+	let issue_html_url = issue.html_url.as_ref().context(error::MissingData)?;
+	let project_html_url =
+		project.html_url.as_ref().context(error::MissingData)?;
+
+	let confirmed_matches_last = db_entry
+		.last_confirmed_project
+		.map(|proj_id| proj_id == confirmed_id)
+		.unwrap_or(false);
+
+	if !confirmed_matches_last {
+		db_entry.issue_confirm_project_ping = None;
+		db_entry.last_confirmed_project = Some(confirmed_id);
+	}
+
+	if project_column.id != confirmed_id {
+		// project has been changed since
+		// the confirmation
+		db_entry.issue_confirm_project_ping = Some(SystemTime::now());
+		db_entry.issue_project_state =
+			Some(ProjectState::Unconfirmed(project_column.id));
+		if let Some(room_id) = &project_info.matrix_room_id {
+			matrix_bot.send_public_message(
+				&room_id,
+				&ISSUE_CONFIRM_PROJECT_MESSAGE
+					.replace("{issue_url}", issue_html_url)
+					.replace("{project_url}", project_html_url)
+					.replace("{issue_id}", &format!("{}", issue_id))
+					.replace(
+						"{project_column.id}",
+						&format!("{}", project_column.id),
+					),
+			)?;
+		} else {
+			// project info should include matrix
+			// room id. TODO some kind of
+			// notification here
+		}
+	}
+	Ok(DbEntryState::Update)
+}
+
+pub fn handle_issue(
+	db: &DB,
+	github_bot: &GithubBot,
+	matrix_bot: &MatrixBot,
+	core_devs: &[github::User],
+	github_to_matrix: &HashMap<String, String>,
+	projects: Option<&Vec<(github::Project, project_info::ProjectInfo)>>,
+	issue: &github::Issue,
+	default_channel_id: &str,
+) -> Result<()> {
+	// TODO: handle multiple projects in a single repo
+
+	let issue_id = issue.id.context(error::MissingData)?;
+
+	let db_key = format!("{}", issue_id).into_bytes();
+	let mut db_entry = DbEntry::new_or_with_key(db, db_key);
+
+	let author_is_core =
+		core_devs.iter().find(|u| u.id == issue.user.id).is_some();
+
+	match if projects.map_or(true, |p| p.is_empty()) {
```

Comment:
Having a `match` around a very complicated `if` statement is huge code smell. This should be broken up with `let`s.